replication/AmbryReplication.tla

```diff
@@ -0,0 +1,328 @@
+---------------------- MODULE AmbryReplication ---------------------
+
+EXTENDS Naturals
+(***************************************************************************)
+(* This specification describes the replication in Ambry, in which         *)
+(* messages can come to replicas of an Ambry partition. The messages can   *)
+(* be "PUT", "DELETE", "TTL UPDATE" (undelete will be covered later).      *)
+(***************************************************************************)
+
+
+(* KNOWN ISSUES IN THE MODEL: (fixes are WIP) *)
+(* The sucess target of put, delete and update-ttl operations is not configurable *)
+(* Replication replicates everything in each iteration. It should rather replicate only those blobs that need replication *)
+(* Stored messages per replica don't have ordering. Hence replication is also not ordered. *)
+(* No concept of replica token. *)
+(* All the messages are replicated in one shot. *)
+(* No distinction between local and remote replicas. *)
+(* Update ttl should have an expiration time and replicas should handle that. *)
+
+(* TODO
+1. MAKE THE REPLICA STORED MESSAGES AN ORDERED LOG
+2. MAKE REPLICATION HAPPEN VIA A REPLICATOKEN SO THAT ONLY NEW MESSAGES ARE REPLICATED
+3. CREATE THE CONCEPT OF EXPIRATION TIME OF A BLOB.
+*)
+
+
+CONSTANT NumBlobs \* Total number of blob messages in a replica.
+CONSTANT NumReplicas \* Total number of replicas of a partition.
+
+(* BlobIds is the set of all blobids that can be present in Ambry.
+ * Ambry generates the blobid for each user requests.
+ * In this spec, everytime new blob is written to Ambry, ambry increments a zero-based counter and uses that as the blobid.
+ *)
+BlobIds == 0 .. NumBlobs-1
+
+Replicas == 0 .. NumReplicas-1 \* Replicas are the set of all replicas of a partition present in Ambry.
+
+(* For the ReplicaStates, *)
+(*     "CanTakeWrite" represents "Leader", "Standby" & "Bootstrapping" states *)
+(*     "CannotTakeWrite" represents "Offline" & "Inactive" states *)
+CanTakeWrite == "CanTakeWrite" \* represents "Leader", "Standby" & "Bootstrapping" states
+CannotTakeWrite == "CannotTakeWrite" \* represents "Offline" & "Inactive" states
+
+PUT == "Put" \* Represents PUT message i.e, request to create a blob.
+UPDATE_TTL == "UpdateTtl" \* Represents an update ttl message. In Ambry, update ttl just makes a blob permanent.
+DELETE == "Delete" \* Represents a request to delete a blob.
+
+ReplicaStateSet == {CanTakeWrite, CannotTakeWrite}
+BlobOpSet == {PUT, DELETE, UPDATE_TTL}
+
+
+ASSUME NumBlobsIsPosNat == NumBlobs \in Nat
+ASSUME NumReplicasIsPosNat == NumReplicas \in Nat
+
+
+VARIABLES
+    replicaStates, \* State of each replica.
+    storedMessagesPerReplica, \* Set of all blob messages present in a replica.
+    blobIdCtr, \* Counter to represent blobid. When a new blob has to be Put, this counter will be used as the blob's blobid.
```

Comment:
Does the platform in real life rely on a counter for assigning unique blob Id's, or is this meant as a simplification for a UUID assignment?  If the latter, maybe make that clear in the comments, if the former, is it relevant to this model to allow for decentralized statefulness of this counter? Or is there some central authority at play?

Reply:
The platform doesn't rely on a counter for assigning blob ids. The assignment of blobid is completely local with no shared state anywhere. So the above logic is meant as a simplification for UUID assignment. I will update the comments.